File: src/programmers/PRO_L2_쇠막대기.cpp
/* 쇠막대기
 * 2019.10.25
 */
#include <iostream>
#include <stack>
#include <string>

using namespace std;

int solution(string arrangement) {
    
    stack<char> s;
    int bar = 0;
    for (int i = 0; i < arrangement.length(); i++) {
        if (arrangement[i] == '(') {
            s.push(arrangement[i]);
            if (arrangement[i+1] == ')') {
                s.pop();
                bar += s.size();
                i += 1;
            }
        } else if (arrangement[i] == ')') {
            s.pop();
            bar += 1;
        }
    }
    int answer = bar;
    return answer;
}

int main() {
    string arrangement = "()(((()())(())()))(())";
    cout << solution(arrangement); // 17
    return 0;
}

File: src/programmers/PRO_L2_쇠막대기.java
package programmers;
class PRO_L2_쇠막대기 {
    public int solution(String arrangement) {
        int answer = 0;
        int cnt = 0;
        for (int i = 0; i < arrangement.length(); i++) {
        	if (arrangement.charAt(i) == '(') {
        		if (arrangement.charAt(i+1) == ')') {
            		answer += cnt;
            		i++;
        		} else {
        			cnt++;
        		}
        	} else {
        		cnt--;
        		answer++;
        	}
        }
        return answer;
    }
}